### PR TITLE
Initialise nodal, multimaterial properties

### PR DESCRIPTION
**Pull request by thiagordonho**
Implement the initialisation of all the properties in the nodal properties pool and call this function from the `MPMExplicit`. The following describe the rational behind the `initialise_properties` function:

- Iterates over all properties in the `std::map` of properties.
- Creates a null matrix of the same size of the property's `Eigen::Matrix`
- Assign the null matrix to the entire property's `Eigen::Matrix`using the previously implemented `assign_property` in `NodalProperties`.

After proper implementation, call `initialise_properties()` at the beginning of the computational step loop.

This PR is related to the RFC #636 .
A reset to `zero` of all the properties' values is expected with the changes in this PR. This initialisation is necessary in order to recalculate the multimaterial properties with every step in the Contact Algorithm.

### CODE REVIEW

File: include/mesh.tcc [hunk 1]
```diff
@@ -1909,3 +1909,10 @@ void mpm::Mesh<Tdim>::create_nodal_properties() {
     throw std::runtime_error("Number of nodes or number of materials is zero");
   }
 }
+
+// Initialise the nodal properties' map
+template <unsigned Tdim>
+void mpm::Mesh<Tdim>::initialise_properties() {
```
Comment (kks32) [inline include/mesh.tcc]: Rename to initialise_nodal_properties

File: include/solvers/mpm_explicit.tcc [hunk 2]
```diff
@@ -170,11 +170,15 @@ bool mpm::MPMExplicit<Tdim>::solve() {
...
     task_group.wait();
 
-    // Assign material ids to node
-    if (interface_)
+    // Initialise nodal properties and assign material ids to node
+    if (interface_) {
+      // Initialise nodal properties
+      mesh_->initialise_properties();
+
       mesh_->iterate_over_particles(
```
Comment (kks32) [inline include/solvers/mpm_explicit.tcc:178]: Does this reset the material ids already present in the node?

File: src/nodal_properties.cc [hunk 3]
```diff
@@ -42,3 +43,17 @@ void mpm::NodalProperties::update_property(
   properties_.at(property).block(node_id * nprops, mat_id, nprops, 1) =
       property_value + this->property(property, node_id, mat_id, nprops);
 }
+
+// Initialise all the nodal values for all properties in the property pool
+void mpm::NodalProperties::initialise_properties() {
+  // Iterate over all properties in the property map
+  for (std::map<std::string, Eigen::MatrixXd>::iterator prop_itr =
```
Comment (kks32) [inline src/nodal_properties.cc]: use auto

File: src/nodal_properties.cc [hunk 4]
```diff
@@ -42,3 +43,17 @@ void mpm::NodalProperties::update_property(
...
+           properties_.begin();
+       prop_itr != properties_.end(); ++prop_itr) {
+    // Create Matrix with zero values that has same size of the current property
+    // in the iteration
+    Eigen::MatrixXd zeroed_property =
```
Comment (kks32) [inline src/nodal_properties.cc:57]: Is the size the entire size of particles * # of materials?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/solvers/mpm_explicit.tcc (hunk 2 above)
Comment (thiagordonho) [reply include/solvers/mpm_explicit.tcc:178]: It does not. It only resets the nodal properties' values. But now that you have said it, it should also reset the material ids in the node as well, right?

File: src/nodal_properties.cc (hunk 4 above)
Comment (thiagordonho) [reply src/nodal_properties.cc:57]: The size is equal to `rows * cols` where:
- `rows` = the number of nodes * property dimension (1 if scalar, 2 or 3 if 2D or 3D vector, respectively)
- `cols` = number of materials

**Comment by thiagordonho** [comment]
@kks32 , how can I fix this code coverage issue?

File: include/solvers/mpm_explicit.tcc (hunk 2 above)
Comment (thiagordonho) [reply include/solvers/mpm_explicit.tcc:178]: material_ids_.clear() is called in the Node class (`initialise()` function), thus already being reset as the nodes are reset.

**Review by kks32 — APPROVED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**